src/Npgsql/TypeHandling/NpgsqlTypeHandler.cs

```diff
@@ -121,10 +131,51 @@ internal virtual ValueTask<object> ReadPsvAsObject(NpgsqlReadBuffer buf, int len
         /// </summary>
         protected internal abstract int ValidateAndGetLength<TAny>([CanBeNull] TAny value, ref NpgsqlLengthCache lengthCache, NpgsqlParameter parameter);
 
+        internal int ValidateAndGetLengthEntry<TAny>(TAny value, ref NpgsqlLengthCache lengthCache, NpgsqlParameter parameter)
+            => NullableHandler<TAny>.Exists
+                ? NullableHandler<TAny>.Validate(this, value, ref lengthCache, parameter)
+                : ValidateAndGetLength(value, ref lengthCache, parameter);
+
         /// <summary>
-        /// Called to write the value of a generic <see cref="NpgsqlParameter{T}"/>.
+        /// Typically does not need to be overridden by type handlers, but may be needed in some
+        /// cases (e.g. <see cref="ArrayHandler"/>.
+        /// Note that this method assumes it can write 4 bytes of length (already verified by
+        /// <see cref="WriteWithLengthEntry{TAny}"/>).
         /// </summary>
-        internal abstract Task WriteWithLengthInternal<TAny>([CanBeNull] TAny value, NpgsqlWriteBuffer buf, NpgsqlLengthCache lengthCache, NpgsqlParameter parameter, bool async);
+        protected internal abstract Task WriteWithLength<TAny>([CanBeNull] TAny value, NpgsqlWriteBuffer buf, NpgsqlLengthCache lengthCache, NpgsqlParameter parameter, bool async);
+
+        internal Task WriteWithLengthEntry<TAny>(TAny value, NpgsqlWriteBuffer buf, NpgsqlLengthCache lengthCache, NpgsqlParameter parameter, bool async)
```

Comment:
It's nice that you lifted this from `NpgsqlTypeHandler<T>` to `NpgsqlTypeHandler`.